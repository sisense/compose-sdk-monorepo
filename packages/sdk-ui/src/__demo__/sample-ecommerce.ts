import { measureFactory } from '@sisense/sdk-data';

import { Commerce, Brand, Country, Category, DataSource } from './sample-ecommerce-autogenerated';

export { Commerce, Brand, Category, Country, DataSource };

/** adding predefined set of measures */
export const Measures = {
  SumCost: measureFactory.sum(Commerce.Cost).format('0,0$'),
  AvgCost: measureFactory.average(Commerce.Cost).format('0,0$'),
  SumRevenue: measureFactory.sum(Commerce.Revenue, 'Total Revenue').format('0,0$'),
  Quantity: measureFactory.sum(Commerce.Quantity, 'Total Quantity').format('0,0'),
  CountBrands: measureFactory.countDistinct(Commerce.BrandID).format('0,0'),
};
