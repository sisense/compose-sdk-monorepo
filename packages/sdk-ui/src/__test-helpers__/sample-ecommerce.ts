import {
  Dimension,
  DateDimension,
  Attribute,
  createAttribute,
  createDateDimension,
  createDimension,
} from '@sisense/sdk-data';

export const DataSource = 'Sample ECommerce';

interface BrandDimension extends Dimension {
  Brand: Attribute;
  BrandID: Attribute;
}
export const Brand = createDimension({
  name: 'Brand',
  Brand: createAttribute({
    name: 'Brand',
    type: 'text-attribute',
    expression: '[Brand.Brand]',
  }),
  BrandID: createAttribute({
    name: 'BrandID',
    type: 'numeric-attribute',
    expression: '[Brand.Brand ID]',
  }),
}) as BrandDimension;

interface CategoryDimension extends Dimension {
  Category: Attribute;
  CategoryID: Attribute;
}
export const Category = createDimension({
  name: 'Category',
  Category: createAttribute({
    name: 'Category',
    type: 'text-attribute',
    expression: '[Category.Category]',
  }),
  CategoryID: createAttribute({
    name: 'CategoryID',
    type: 'numeric-attribute',
    expression: '[Category.Category ID]',
  }),
}) as CategoryDimension;

interface CommerceDimension extends Dimension {
  AgeRange: Attribute;
  BrandID: Attribute;
  CategoryID: Attribute;
  Condition: Attribute;
  Cost: Attribute;
  CountryID: Attribute;
  DateMonth: Attribute;
  Gender: Attribute;
  Quantity: Attribute;
  Revenue: Attribute;
  VisitID: Attribute;
  Date: DateDimension;
}
export const Commerce = createDimension({
  name: 'Commerce',
  AgeRange: createAttribute({
    name: 'AgeRange',
    type: 'text-attribute',
    expression: '[Commerce.Age Range]',
  }),
  BrandID: createAttribute({
    name: 'BrandID',
    type: 'numeric-attribute',
    expression: '[Commerce.Brand ID]',
  }),
  CategoryID: createAttribute({
    name: 'CategoryID',
    type: 'numeric-attribute',
    expression: '[Commerce.Category ID]',
  }),
  Condition: createAttribute({
    name: 'Condition',
    type: 'text-attribute',
    expression: '[Commerce.Condition]',
  }),
  Cost: createAttribute({
    name: 'Cost',
    type: 'numeric-attribute',
    expression: '[Commerce.Cost]',
  }),
  CountryID: createAttribute({
    name: 'CountryID',
    type: 'numeric-attribute',
    expression: '[Commerce.Country ID]',
  }),
  DateMonth: createAttribute({
    name: 'DateMonth',
    type: 'numeric-attribute',
    expression: '[Commerce.Date (Month)]',
  }),
  Gender: createAttribute({
    name: 'Gender',
    type: 'text-attribute',
    expression: '[Commerce.Gender]',
  }),
  Quantity: createAttribute({
    name: 'Quantity',
    type: 'numeric-attribute',
    expression: '[Commerce.Quantity]',
  }),
  Revenue: createAttribute({
    name: 'Revenue',
    type: 'numeric-attribute',
    expression: '[Commerce.Revenue]',
  }),
  VisitID: createAttribute({
    name: 'VisitID',
    type: 'numeric-attribute',
    expression: '[Commerce.Visit ID]',
  }),
  Date: createDateDimension({
    name: 'Date',
    expression: '[Commerce.Date (Calendar)]',
  }),
}) as CommerceDimension;

interface CountryDimension extends Dimension {
  Country: Attribute;
  CountryID: Attribute;
}
export const Country = createDimension({
  name: 'Country',
  Country: createAttribute({
    name: 'Country',
    type: 'text-attribute',
    expression: '[Country.Country]',
  }),
  CountryID: createAttribute({
    name: 'CountryID',
    type: 'numeric-attribute',
    expression: '[Country.Country ID]',
  }),
}) as CountryDimension;
