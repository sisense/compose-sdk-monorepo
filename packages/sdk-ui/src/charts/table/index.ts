export * from './pure_table';
export type { TableProps } from './types';
export type { TableDesignOptions } from '../../chart-options-processor/translations/design_options';
