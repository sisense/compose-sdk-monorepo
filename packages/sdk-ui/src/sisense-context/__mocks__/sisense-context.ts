export const useSisenseContext = vi.fn();
export const useSisenseContextMock = useSisenseContext;
