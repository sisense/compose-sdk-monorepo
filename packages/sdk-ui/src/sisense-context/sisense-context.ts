import { createContext, useContext } from 'react';
import { ClientApplication } from '../app/client-application';

export type SisenseContextPayload = {
  isInitialized: boolean;
  app?: ClientApplication;
  tracking: {
    enabled: boolean;
    packageName: string;
  };
};

export const SisenseContext = createContext<SisenseContextPayload>({
  isInitialized: false,
  tracking: {
    enabled: true,
    packageName: 'sdk-ui',
  },
});

export const useSisenseContext = () => useContext(SisenseContext);
