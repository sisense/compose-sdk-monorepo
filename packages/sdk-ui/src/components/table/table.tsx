/* eslint-disable max-lines-per-function */
import React, { useEffect, useRef, useState } from 'react';
import { Pagination } from '@mui/material';
import { isDataSource } from '@sisense/sdk-data';
import { TableProps } from '../../props';
import { useThemeContext } from '../theme-provider';
import { translateTableDataOptions } from '../../chart-data-options/translate-data-options';
import { useTableData } from './hooks/use-table-data';
import { useTableDataTable } from './hooks/use-table-datatable';
import { Column as DataTableColumn } from '../../chart-data-processor/table-processor';
import { updateInnerDataOptionsSort } from '../../chart-data/table-data';
import { PureTable, TableDesignOptions } from '../../charts/table';
import { getCustomPaginationStyles } from './styles/get-custom-pagination-styles';
import { generateUniqueDataColumnsNames } from '../../chart-data-options/validate-data-options';
import { isValue, TableDataOptionsInternal } from '../../chart-data-options/types';
import { isDataTableEmpty } from '../../chart-data-processor/table-creators';
import { NoResultsOverlay } from '../no-results-overlay/no-results-overlay';
import { asSisenseComponent } from '../decorators/as-sisense-component';
import { shouldSkipSisenseContextWaiting } from '../chart';
import { DynamicSizeContainer, getChartDefaultSize } from '../dynamic-size-container';

/**
 * Table with aggregation and pagination.
 *
 * @example
 * (1) Example of Table of raw data from the `Sample ECommerce` data model:
 *
 * ```tsx
 * <Table
 *   dataSet={DM.DataSource}
 *   dataOptions={{
 *     columns: [
 *       DM.Commerce.AgeRange,
 *       DM.Commerce.Revenue,
 *       DM.Commerce.Cost,
 *       DM.Commerce.Quantity,
 *     ],
 *   }}
 *   styleOptions={{ width: 600, height: 750 }}
 * />
 * ```
 * ###
 * <img src="media://table-example-1.png" width="800px" />
 *
 * (2) Example of Table of aggregated data from the same data model:
 *
 * ```tsx
 * <Table
 *   dataSet={DM.DataSource}
 *   dataOptions={{
 *     columns: [
 *       DM.Commerce.AgeRange,
 *       measures.sum(DM.Commerce.Revenue, 'Total Revenue'),
 *       measures.sum(DM.Commerce.Cost, 'Total Cost'),
 *     ],
 *   }}
 *   styleOptions={{
 *     headersColor: true,
 *     alternatingColumnsColor: false,
 *     alternatingRowsColor: true,
 *   }}
 * />
 * ```
 * ###
 * <img src="media://table-example-2.png" width="800px" />
 * @param props - Table properties
 * @returns Table component
 */

export const Table = asSisenseComponent({
  componentName: 'Table',
  shouldSkipSisenseContextWaiting,
})(({ dataSet, dataOptions, styleOptions = {}, filters }: TableProps) => {
  const { rowsPerPage = 25, width, height } = styleOptions;
  const pagesCountLoadingStep = 10;
  const { themeSettings } = useThemeContext();
  const [innerDataOptions, setInnerDataOptions] = useState<TableDataOptionsInternal | null>(null);
  const [dataColumnNamesMapping, setDataColumnNamesMapping] = useState({});
  const [offset, setOffset] = useState(0);
  const data = useTableData({
    dataSet,
    dataOptions: innerDataOptions,
    filters,
    count: rowsPerPage * pagesCountLoadingStep,
    offset,
  });
  const [currentPage, setCurrentPage] = useState(1);
  const paginationEl = useRef(null);

  const dataTable = useTableDataTable({
    data,
    innerDataOptions,
    dataColumnNamesMapping,
    needToAggregate: dataSet !== undefined && !isDataSource(dataSet),
  });

  useEffect(() => {
    const translatedDataOptions = translateTableDataOptions(dataOptions);
    setDataColumnNamesMapping(
      generateUniqueDataColumnsNames(translatedDataOptions.columns.filter(isValue)),
    );
    setInnerDataOptions(translatedDataOptions);

    setOffset(0);
  }, [dataOptions]);

  const onPageChange = (page: number) => {
    if (!data) return;
    const lastPage = Math.floor(data.rows.length / rowsPerPage) === page;
    if (lastPage) {
      setOffset(data.rows.length);
    }
    setCurrentPage(page);
  };

  const onSortUpdate = (column: DataTableColumn) => {
    setCurrentPage(1);
    setOffset(0);
    if (innerDataOptions) {
      setInnerDataOptions(updateInnerDataOptionsSort(innerDataOptions, column));
    }
  };

  if (!dataTable || !innerDataOptions) return null;

  const paginatedTable = {
    columns: dataTable.columns,
    rows: dataTable.rows.slice((currentPage - 1) * rowsPerPage, currentPage * rowsPerPage),
  };
  const pagesCount = Math.ceil(dataTable.rows.length / rowsPerPage);
  const paginationHeight = pagesCount > 1 ? 32 : 0;

  return (
    <DynamicSizeContainer
      defaultSize={getChartDefaultSize('table')}
      size={{
        width,
        height,
      }}
    >
      {(size) => {
        if (isDataTableEmpty(dataTable)) {
          return <NoResultsOverlay iconType={'table'} />;
        }

        return (
          <div
            style={{
              display: 'flex',
              flexDirection: 'column',
              alignItems: 'center',
              backgroundColor: themeSettings.chart.backgroundColor,
            }}
            aria-label="table-root"
          >
            <PureTable
              dataTable={paginatedTable}
              dataOptions={innerDataOptions}
              designOptions={styleOptions as TableDesignOptions}
              themeSettings={themeSettings}
              width={size.width}
              height={size.height - paginationHeight}
              onSortUpdate={onSortUpdate}
            />
            {pagesCount > 1 && (
              <Pagination
                ref={paginationEl}
                page={currentPage}
                count={pagesCount}
                onChange={(event, page) => onPageChange(page)}
                sx={getCustomPaginationStyles(themeSettings)}
              />
            )}
          </div>
        );
      }}
    </DynamicSizeContainer>
  );
});
