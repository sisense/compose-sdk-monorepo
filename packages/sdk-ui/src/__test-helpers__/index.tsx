/**
 * Common test utilities
 */
import { PropsWithChildren, ReactElement } from 'react';
import { render, RenderResult } from '@testing-library/react';
import type { Cell, Data } from '@ethings-os/sdk-data';
import isObject from 'lodash-es/isObject';
import userEvent from '@testing-library/user-event';
import { SisenseContext, SisenseContextPayload } from '../sisense-context/sisense-context';
import { Authenticator, HttpClient } from '@ethings-os/sdk-rest-client';
import { ClientApplication } from '../app/client-application';
import { DeepPartial } from 'ts-essentials';
import { useTranslation } from 'react-i18next';
import { translation } from '../translation/resources/en';
import { CustomWidgetsProvider } from '../custom-widgets-provider';
import { MenuProvider } from '@/common/components/menu/menu-provider';

type UserSetupFn = (typeof userEvent)['setup'];
interface SetupResult extends RenderResult {
  user: ReturnType<UserSetupFn>;
}
/**
 * This helper method combines the userEvent.setup() and render() calls at the
 * start of a unit test using @testing-library/react.
 *
 * Based on recommendation by: https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent
 *
 * @param jsx - React component instance to test, passed directly to render()
 * @param useFakeTimers - true if Vitest fake timers are enabled
 * @returns
 */
export function setup(jsx: ReactElement, useFakeTimers = false): SetupResult {
  return {
    // Need to set "advanceTimers" if we're using Vitest's fake timers.
    // The solution in https://github.com/testing-library/user-event/issues/833#issuecomment-1171452841
    // is for Jest, but this seems to work for Vitest as well.
    user: userEvent.setup(
      useFakeTimers ? { advanceTimers: (ms) => vi.advanceTimersByTime(ms) } : undefined,
    ),
    ...render(jsx),
  };
}

export const MockedHighchartsReact = ({
  options,
}: {
  /**
   * Highcharts chart configuration object.
   * Please refer to the Highcharts (API documentation)[https://api.highcharts.com/highcharts/].
   */
  options: object;
}) => {
  return <div>{JSON.stringify(options)}</div>;
};

type MockedSisenseContextProviderProps = PropsWithChildren<
  DeepPartial<Omit<SisenseContextPayload, 'app'>>
>;

const mockHttpClient = new HttpClient(
  'http://mock-url/sometenant?someparam=true',
  {} as Authenticator,
  'mock-env',
);
export const MockedSisenseContextProvider = ({
  children,
  tracking = {
    enabled: true,
  },
  isInitialized = true,
  errorBoundary,
}: MockedSisenseContextProviderProps) => {
  return (
    <SisenseContext.Provider
      value={{
        app: {
          httpClient: mockHttpClient,
        } as ClientApplication,
        tracking: {
          enabled: tracking.enabled ?? true,
          packageName: tracking.packageName ?? 'sdk-ui',
        },
        isInitialized,
        errorBoundary: { showErrorBox: true, ...errorBoundary },
      }}
    >
      <MenuProvider>
        <CustomWidgetsProvider>{children}</CustomWidgetsProvider>
      </MenuProvider>
    </SisenseContext.Provider>
  );
};

/**
 * Extends the provided data by blurring specified rows.
 *
 * @param {Data} data - The original data object.
 * @param {number[]} blurRowsIndexes - An array of row indexes to be blurred.
 * @returns {Data} The data object with blurred rows.
 */
export const withBlurredRows = (data: Data, blurRowsIndexes: number[]) => ({
  ...data,
  rows: data.rows.map((row, rowIndex) =>
    row.map((value) => {
      const cell: Cell = isObject(value) ? value : { data: value };

      if (blurRowsIndexes.includes(rowIndex)) {
        return { ...cell, blur: true };
      }
      return cell;
    }),
  ),
});

/**
 * Mocking the i18next translation functionality for tests.
 */
export const setupI18nMock = () => {
  vi.mock('react-i18next', async () => {
    const original: { useTranslation: typeof useTranslation } = await vi.importActual(
      'react-i18next',
    );

    // Helper function to get nested translations safely
    const getNestedTranslation = (key: string): string | undefined => {
      return key.split('.').reduce((accumulator: any, currentKey: string) => {
        if (accumulator && typeof accumulator === 'object' && currentKey in accumulator) {
          return accumulator[currentKey];
        }
        return undefined;
      }, translation) as string | undefined;
    };

    return {
      ...original,
      useTranslation: () => ({
        ...original.useTranslation(),
        t: vi.fn((key, options?: Record<string, any>) => {
          const translatedValue = getNestedTranslation(key);
          if (translatedValue && options) {
            // Replace variables in the translated string
            return translatedValue.replace(/{{(.*?)}}/g, (_, varName) => {
              const trimmedVarName = varName.trim();
              return options[trimmedVarName] || `{{${trimmedVarName}}}`; // Fallback if variable is not found
            });
          }
          return translatedValue ?? key; // Fallback to the key if translation is not found
        }),
      }),
    };
  });
};
