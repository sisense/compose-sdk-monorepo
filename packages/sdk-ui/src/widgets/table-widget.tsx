/* eslint-disable @typescript-eslint/naming-convention */
import React, { type FunctionComponent, useState } from 'react';

import { getDataSourceName } from '@sisense/sdk-data';

import { useSisenseContext } from '@/sisense-context/sisense-context';
import { Table } from '@/table';

import { asSisenseComponent } from '../decorators/component-decorators/as-sisense-component';
import { DynamicSizeContainer, getWidgetDefaultSize } from '../dynamic-size-container';
import { TableWidgetProps } from '../props';
import { WidgetContainer } from './common/widget-container';

/**
 * The TableWidget component extending the Table component to support widget style options.
 *
 * @example
 * Example of using the `Widget` component to
 * plot a bar chart of the `Sample ECommerce` data source hosted in a Sisense instance.
 * Drill-down capability is enabled.
 * ```tsx
 * <TableWidget
 *   dataSource={DM.DataSource}
 *   dataOptions={{
 *     columns: [DM.Category.Category]
 *   }}
 * />
 * ```
 * @param props - Table Widget properties
 * @returns Widget component representing a table
 * @internal
 */
export const TableWidget: FunctionComponent<TableWidgetProps> = asSisenseComponent({
  componentName: 'TableWidget',
})((props) => {
  const [refreshCounter, setRefreshCounter] = useState(0);
  const { app } = useSisenseContext();

  const { styleOptions, dataSource = app?.defaultDataSource, dataOptions } = props;

  const defaultSize = getWidgetDefaultSize('table', {
    hasHeader: !styleOptions?.header?.hidden,
  });
  const { width, height, ...styleOptionsWithoutSizing } = props.styleOptions || {};

  if (!dataOptions) {
    return null;
  }

  return (
    <DynamicSizeContainer
      defaultSize={defaultSize}
      size={{
        width: width,
        height: height,
      }}
    >
      <WidgetContainer
        {...props}
        dataSetName={dataSource && getDataSourceName(dataSource)}
        onRefresh={() => setRefreshCounter(refreshCounter + 1)}
      >
        <Table
          dataSet={props.dataSource}
          dataOptions={props.dataOptions}
          styleOptions={styleOptionsWithoutSizing}
          filters={props.filters}
          refreshCounter={refreshCounter}
        />
      </WidgetContainer>
    </DynamicSizeContainer>
  );
});
