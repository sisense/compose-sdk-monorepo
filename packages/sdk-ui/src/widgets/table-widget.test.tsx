import { render } from '@testing-library/react';
import { TableWidget } from './table-widget';
import { Table } from '@/table';
import { filterFactory } from '@sisense/sdk-data';
import { TableDataOptions } from '..';
import { useSisenseContext } from '@/sisense-context/sisense-context';
import * as DM from '@/__test-helpers__/sample-ecommerce';
import { Mock } from 'vitest';

vi.mock('@/sisense-context/sisense-context');
vi.mock('@/table', () => ({
  Table: vi.fn(() => <div>Mocked Table</div>),
}));

describe('TableWidget Component', () => {
  const mockDataOptions: TableDataOptions = { columns: [] };
  const mockStyleOptions = { width: 500, height: 300, header: { hidden: false } };

  beforeEach(() => {
    vi.clearAllMocks();
    // Mock useSisenseContext to return the expected shape
    (useSisenseContext as Mock).mockReturnValue({
      isInitialized: true,
      app: {
        defaultDataSource: DM.DataSource,
      },
      errorBoundary: {
        showErrorBox: true,
      },
    });
  });

  it('renders Table with correct props', () => {
    const { getByText } = render(
      <TableWidget
        dataSource={DM.DataSource}
        dataOptions={mockDataOptions}
        styleOptions={mockStyleOptions}
        filters={[
          filterFactory.members(DM.Commerce.Condition, ['Used', 'Refurbished'], {
            guid: 'test-id',
          }),
        ]}
      />,
    );

    expect(getByText('Mocked Table')).toBeInTheDocument();
    const tableProps = (Table as Mock).mock.calls[0][0];
    expect(tableProps).toMatchSnapshot();
  });

  it('does not render Table when dataOptions are not provided', () => {
    const { queryByText } = render(
      <TableWidget
        styleOptions={mockStyleOptions}
        dataOptions={undefined as unknown as TableDataOptions}
      />,
    );
    expect(queryByText('Mocked Table')).not.toBeInTheDocument();
  });
});
