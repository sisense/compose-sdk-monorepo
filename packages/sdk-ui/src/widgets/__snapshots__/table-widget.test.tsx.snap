// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableWidget Component > renders Table with correct props 1`] = `
{
  "dataOptions": {
    "columns": [],
  },
  "dataSet": "Sample ECommerce",
  "filters": [
    {
      "__serializable": "DimensionalElement",
      "attribute": {
        "__serializable": "DimensionalElement",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "desc": "",
        "expression": "[Commerce.Condition]",
        "name": "Condition",
        "type": "text-attribute",
      },
      "dataSource": undefined,
      "desc": "",
      "filterType": "members",
      "members": [
        "Used",
        "Refurbished",
      ],
      "name": "28400200b58",
      "type": "filter",
    },
  ],
  "refreshCounter": 0,
  "styleOptions": {
    "header": {
      "hidden": false,
    },
  },
}
`;
